src/Components/Components/src/RenderTree/EventHandlerDelegateWithClosuresComparer.cs

```diff
@@ -0,0 +1,130 @@
+        {
+            if (Equals(left, right))
+            {
+                return true;
+            }
+
+            // If any of the delegates or their targets are null (static delegates) then normal
+            // equality is sufficient. No additional testing required.
+            if (left == null || right == null || left.Target == null || right.Target == null)
+            {
+                return false;
+            }
+
+            var oldTargetType = left.Target!.GetType();
+            var newTargetType = right.Target!.GetType();
+
+            // if the types are not the same, or the methods are not the same, these possible closures are
+            // not the same anyway.
+            if (oldTargetType != newTargetType || left.Method != right.Method)
+            {
+                return false;
+            }
+
+            // for speed we cache the comparison functions per closure type.
+            var comparison = ClosureComparerCache.GetOrAdd(newTargetType, GetClosureComparerForType);
+            return comparison(left.Target, right.Target);
+        }
+
+        private static Func<object, object, bool> GetClosureComparerForType([DynamicallyAccessedMembers(DynamicallyAccessedMemberTypes.PublicFields)] Type type)
+        {
+            // if this type is not compiler generated, it is not a closure, and the previous tests where
+            // enough to conclude that these are not equal. We just cache a delegate to a function that returns false.
+            if (!type.CustomAttributes.Any(x => x.AttributeType == typeof(CompilerGeneratedAttribute)))
+            {
+                return (objA, objB) => false;
+            }
+
+            // build an array of lambdas that will get each public field value on the closure.
+            var fieldGetters = type.GetFields(
+                BindingFlags.Public |
+                BindingFlags.Instance |
+                BindingFlags.DeclaredOnly).Select(fieldInfo =>
+            {
+                var sourceParam = Expression.Parameter(typeof(object));
+                Expression returnExpression = Expression.Field(Expression.Convert(sourceParam, fieldInfo.DeclaringType!), fieldInfo);
+                if (fieldInfo.FieldType.IsValueType)
+                {
+                    // box if neccessary.
+                    returnExpression = Expression.Convert(returnExpression, typeof(object));
+                }
+                var lambda = Expression.Lambda(returnExpression, sourceParam);
+                return (Func<object?, object?>)lambda.Compile();
+            }).ToArray();
+
+            // the actual comparison function. It re-uses the array of public field 'getters' that we have built up
+            // before.
+            return (objA, objB) =>
+            {
+                // We have two instances of the same compiler generated class. Let's compare all public fields.
+                foreach (var fieldGetter in fieldGetters)
```

Comment:
This could be a `for` loop rather than `foreach`. A `foreach` generates a `try`/`catch` which is an additional cost.

Reply:
Could you point me to some info where foreach would generate a try / catch? I've disassembled the code in question here:

https://gist.github.com/jspuij/94e3684a7a7ce46d8ffe70ad24ad6636

The captured variable is an array (on purpose with ToArray()). The compiler recognizes the array and creates a pretty efficient loop with a normal array indexer (see  `ldelem.ref`). No exception handlers that I could find.

Reply:
Yes, sorry, you're right. There would be a `try`/`catch` if it was a `List<T>` or some other general enumerable, but you're right that with a `T[]` the generated code is very minimal and has no `try`/`catch`.